[SCF] Fixed epilogue predicates in loop pipelining

sjw36 opened the pull request:
The computed loop iteration is zero based, so only check it is less than zero. 
This fixes the case when lower bound is not zero.

ThomasRaoux commented on mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp:
```diff
@@ -671,9 +674,9 @@ LoopPipelinerInternal::emitEpilogue(RewriterBase &rewriter,
     if (dynamicLoop) {
-      // pred = iterI >= lb
+      // pred = iterI < 0
       predicates[i + 1] = rewriter.create<arith::CmpIOp>(
-          loc, arith::CmpIPredicate::sge, iterI, lb);
+          loc, arith::CmpIPredicate::slt, iterI, zero);
```
shouldn't this be `pred = iterI >= 0`? So the predicate is true only when iterI is negative seems odd considering `iterI = total_iters - 1 - i`?

sjw36 replied:
Ah, that's right.
So for:
```
scf.for (lb=3..ub=4, step=2) max_stage = 2                                                        
```
This computes:
```
  total_iterations = (4 - 3) + 2 - 1 / 2 = 1                                                    
  iterI(0) = total_iterations - 1 - 0 = 0                                                       
  newlastIter(0) = iterI(0) * 2 + 3 = 3                                                         
  pred(0) = iterI(0) >= 0 = true                                                                
  iterI(1) = total_iterations - 1 - 1 = -1                                                      
  newlastIter(1) = iterI(1) * 2 + 3 = 1                                                         
  pred(1) = iterI(1) >= 0 = false
```
I'll update now.

ThomasRaoux commented:
can we add a test or modify one to not start from 0 so that this can be tested?

sjw36 commented:
@ThomasRaoux updated tests.

ThomasRaoux commented on mlir/test/Dialect/SCF/loop-pipelining.mlir:
```diff
@@ -802,8 +801,10 @@ func.func @stage_0_value_escape(%A: memref<?xf32>, %result: memref<?xf32>, %ub:
-func.func @dynamic_loop(%A: memref<?xf32>, %result: memref<?xf32>, %lb: index, %ub: index, %step: index) {
+func.func @dynamic_loop(%A: memref<?xf32>, %result: memref<?xf32>, %ub: index) {
   %cf = arith.constant 1.0 : f32
+  %lb = arith.constant 3 : index
+  %step = arith.constant 2 : index
```
why do we want those to be constants?

sjw36 replied:
Just added to make a lower_bound non-zero. The loop is still dynamic with the upper_bound passed in.

ThomasRaoux replied:
wasn't it already a lower_bound non-zero?

sjw36 replied:
It helps demonstrates that the `iterI >= 0` is not related to the lower_bound.

ThomasRaoux replied:
how can it be the lower bound if the lower bound is a function argument?

sjw36 replied:
%lb and %step were removed from the func args.

ThomasRaoux replied:
my question is how does changing %ld to be a constant helps demonstrate that iterI >= 0 predicate is not related to the lower_bound.

sjw36 replied:
I can revert that part (lb, step -> constants) but it should cover the dynamic loop with epilogue peeling either way.

ThomasRaoux replied:
yeah I would remove it. I think it is good to test the general case where nothing is known about the loop bound

Commit:
[SCF] Fixed epilogue predicates in loop pipelining

The computed loop iteration is zero based, so only check it is less than zero.
This fixes the case when lower bound is not zero.

sjw36 commented:
@ThomasRaoux does this look adequate?
